### DISCUSSION TITLE
Obtaining even more powerful information with AST modification like birdseye

### DISCUSSION THREAD

**Discussion by alexmojaki** [aroberge/friendly#128] [Ideas]
Related to the current discussion in #96, just a fun thought of where this could go in the future...

In some cases where you control the execution (e.g. in the friendly console or running a script with `python -m friendly_traceback`) you could modify the AST so that it collects the values of *all* expressions as they are evaluated, still without side effects. This is what is done in https://github.com/alexmojaki/birdseye and https://github.com/alexmojaki/snoop#ppdeep-for-tracing-subexpressions - the former is a good visualisation of what it can do while the latter has a much simpler text-only implementation. It can also identify the node which caused the exception in basically all cases, whereas `executing` is still fairly limited.

I could recreate something similar for you here fairly easily.  However this adds a lot of overhead regardless of whether an exception occurs, so it would need to be optional and used sparingly.

**Comment by aroberge** [comment]
I would rather not do this.  Right now, people can just install friendly-traceback as a sys.excepthook and it just works.  It would be really problematic to have it work in two different ways (modifying or not modifying the AST) and try to deal with two types of situations when an exception is raised.

I've just managed to sort of make it work with IPython (see https://github.com/aroberge/friendly-traceback/issues/126) and I don't believe that modifying the AST would be acceptable in this situation.
